spec/support/aws_utils/orchestrator.rb

```diff
@@ -251,6 +251,56 @@ def provision_auth_ecs_task(public_key_path: nil,
       end
     end
 
+    def wait_for_ecs_ready(
+      cluster_name: AWS_AUTH_ECS_CLUSTER_NAME,
+      service_name: AWS_AUTH_ECS_SERVICE_NAME,
+      timeout: 20
+    )
+      deadline = Time.now + timeout
+
+      # The AWS SDK waiter seems to immediately fail sometimes right after
+      # the service is created, so wait for the service to become active
+      # manually and then use the waiter to wait for the service to become
+      # stable.
+      #
+      # The failure may be due to the fact that apparently, it is possible for
+      # describe_services to not return an existing service for some time.
+      # Therefore, allow the lack of service to be a transient error.
+      loop do
+        service = ecs_client.describe_services(
+          cluster: cluster_name,
+          services: [service_name],
+        ).services.first
+
+        if service.nil?
+          puts "Service #{service_name} in cluster #{cluster_name} does not exist (yet?)"
+          status = 'MISSING'
+        elsif service.status.downcase == 'active'
+          break
+        else
+          status = service.status
+        end
+
+        now = Time.now
+        if now >= deadline
+          raise "Service #{service_name} in cluster #{cluster_name} did not become ready in #{timeout} seconds (current status: #{status})"
+        end
+
+        puts "Wating for service #{service_name} in cluster #{cluster_name} to become ready (#{'%2.1f' % (deadline - now)} seconds remaining, current status: #{status})"
+        sleep 5
+      end
+
+      puts "Wating for service #{service_name} in cluster #{cluster_name} to become stable"
+      ecs_client.wait_until(
+        :services_stable, {
+          cluster: cluster_name,
+          services: [service_name],
+        },
+        delay: 5,
+        max_attempts: 36,
+      )
+    end
+
```

Comment:
Why are we implementing these `aws_utils`? All other drivers (and the server) are using the drivers-evergreen-tools scripts to test MONGODB-AWS.

Reply:
The tooling in the Ruby driver is much more extensive, flexible, modular and integrated than d-e-t. The difference between the two is similar to the difference in testing code locally and sending it to Evergreen for every change, for Evergreen to run the entire test suite and report that there is a syntax error in a file somewhere. It works but is not efficient.

For example, when the driver tooling is used to provision the cluster and task definition, it sets up CloudWatch logs for the task, such that any provisioning errors are can be quickly identified. Subsequently in  cluster so provisioned it is possible to run the tests (following instructions added in this PR under "without rebuilding the remote environment") with about 5 second turnaround time. The minimum turnaround time with Evergreen I've seen is about 3 minutes, and 5-10 minutes is more common if you consider  the number of clicks and time needed to get to the build, the logs, find the problem in the logs, etc.. This is a massive productivity difference when developing.

Thinking back to the issue I ran into with apt prompting for input in the ecs task, I am under the impression that the input prompt was not even shown in the Evergreen logs. Troubleshooting this via Evergreen would've taken an inordinate amount of time or luck. The driver tooling permits running all tests interactively and exposed this issue immediately.

To give another example, the instance role tests perform credential retrieval with the instance profile assigned to the instance as well as when the instance profile is not assigned, and verify that both scenarios work as expected.

Reply:
Thanks for the detailed response. I'm convinced these scripts have value. I hope we can use these to benefit other drivers/implementors in the future.